File: Implementations/Python/Half Spaces-nd.py
# ----------------------------------------------------------
                # Basic Linear Classifier
                # H = sign(<w,x> + b) - Half Spaces
                # L_D(h) - Px~D[h(x) != y]
                # L_S(h) - 0/1 Loss
                # Surrogate Loss - Hinge Loss = max(0,1-(f(x_i)*y)
                # f(x_i) = {-1,1}
                # The Realizeable Case
                # m(e,d) = 
# ----------------------------------------------------------
#%% Imports
import numpy as np                
#%% Data Section
n = 5 # Dimension
ds = 10**4 # Domain Size - (Unknown in real situations)
eps = 0.1 # Requested accuracy of the learner
delta = 2/100 # Probability the learner will fail
vcDim = n+1;
m = int((vcDim + np.log(1/delta))/eps) # Sample Size

def f(x):
    if(np.sum(x) < 1):
        return 1
    else:
        return -1

D = np.random.rand(ds,n)
#%% The Sample Space

S = [D[i] for i in range(m)]
Y = [f(x) for x in S]
#%% Scikit Learn Linear Fit
from sklearn import linear_model
clf = linear_model.SGDClassifier(max_iter=1000, tol=1e-3)
clf.fit(S, Y)
#%%
print("Prediction")      
Risk = 0;
for i in range(len(D)):
    vector = [D[i][j] for j in range(n)]
    if(clf.predict([vector]) != f(D[i])):
        Risk = Risk + 1;
Risk = Risk/len(D)
    
    

File: Implementations/MATLAB/HalfSpacesND.m
% ----------------------------------------------------------
                % Basic Linear Classifier
                % H = sign(<w,x> + b) - Half Spaces
                % L_D(h) - Px~D[h(x) != y]
                % L_S(h) - 0/1 Loss
                % Surrogate Loss - Hinge Loss = max(0,1-(f(x_i)*y)
                % f(x_i) = {-1,1}
                % The Realizeable Case
                % m(e,d) = m(e,d) = VC-Dimension + LOG(1/delta) / epsilon
% ----------------------------------------------------------

%% 
% Let (Omega,S,h,f) be the learning problem such that
% Omega = Real distribution
% S = I.I.D Samples
% h = hypothesis output by the learner
% f = labeling function for the realizeable case
% n = Dimensionality of the distribution
% m = size of sample
% epsilon = requested accuracy of the learner
% delta = maximum probability a learner will fail
n = 5;
eps = 0.1;
delta = 0.02;
vcDim = n+1;
m = ceil((vcDim + log(1/delta))/eps);
ds = 10^8; % Size of the real distribution, unknown in real situations
Omega = rand(ds,n);
realLabels = zeros(1,ds);
for i = 1:ds
    realLabels(i) = labeler(Omega(i,:),3);
end
%%
S = Omega(1:m,1:n); % Sample space sized m with dimension n
labels = realLabels(1:m);

%%
[Mdl,FitInfo] = fitclinear(S,labels);
prediction =  Mdl.predict(Omega);
risk = 0;
for i = 1:ds
    if(prediction(i) ~= realLabels(i))
        risk = risk + 1;
    end
end
risk = risk/ds;